[Merged by Bors] - Ensure that App closes started channel on startup error

piersy opened the pull request:
## Motivation
Closes #4868 

## Changes
This simply moves non blocking app startup operations into a new method allowing for a deferred call to close the started channel. This ensures that started will always be closed before Start returns and thus prevents getting stuck waiting for the `App` to start. There were also a couple of services that were deferring their shutdown to the end of Start, they have now been added as fields of` node.App` and are shutdown in `stopServices` which keeps the shutdown approach consistent.
## Test Plan
Current tests should pass

## DevOps Notes
<!-- Please uncheck these items as applicable to make DevOps aware of changes that may affect releases -->
- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

piersy committed:
Ensure that App closes started channel on startup error

poszu commented on node/node.go:
```diff
@@ -1285,6 +1285,36 @@ func (app *App) setupDBs(ctx context.Context, lg log.Log, dbPath string) error {
+}
+
+func (app *App) startSynchronous(ctx context.Context) error {
+	// notify anyone who might be listening that the app has finished starting.
+	// this can be used by, e.g., app tests.
+	defer close(app.started)
```
Now it will close the channel even in the case of an error during a start. Should it be that way?

piersy replied:
Hey @poszu that is the intention, consider this code:

```go
g.Go(func() error {
	return app1.Start(ctx)
})
<-app1.Started()
```

If there's an error at startup and the started channel is not closed, the caller gets stuck, it doesn't matter that the error is returned through the error group, because the code is not able to progress to the point where it waits on the error group.

poszu replied:
OK, now I understand what you want to do in this PR, actually. But TBH, I don't understand the purpose of the `started` channel. It now synchronizes to the point in time where `startSynchronous()` quits, but how is it useful?

fasmat replied:
Do we even need the channel any more? It is closed when `Start` returns. So if `Start` returns `nil` the start was successful, otherwise it wasn't.

piersy replied:
Hey @poszu & @fasmat the thing is that `Start` runs the app,  so if all goes well `Start` won't return until the app is shutting down. I believe the purpose of the started channel is to let the caller know when it is safe to call the APIs exposed by the app. An alternative would be to separate all listening from serving operations and then app startup could be done in two stages, first setting up all services to listen and a second call to start serving those services. However I think that would be a much larger change, since it would involve separating those stages for many components.

fasmat replied:
Yeah, I momentarily forgot that `app.Start` is a blocking call. Forget my comment about the channel not being needed any more 🙂

poszu replied:
@piersy
> I believe the purpose of the started channel is to let the caller know when it is safe to call the APIs exposed by the app.

But that's not really true. The services are started asynchronously (in goroutines). For example, the GRPC services are started in a goroutine. They might not yet be started when the channel is closed.

piersy replied:
@fasmat Hehe, easy to do, I have done the same :D

@poszu You're right, its a bit of a kludge, you don't actually know for sure that the services are going to be listening after receiving on the channel. However if the channel was removed all you could do is wait and try the API. So I think the channel use is preferable to that, or maybe I've misunderstood the original purpose of this channel, but I can't think what else it's for.

piersy replied:
@poszu I added a ticket for the 2 stage startup. - https://github.com/spacemeshos/go-spacemesh/issues/4892

piersy committed:
Merge branch 'develop' into piersy/fix-app-startup

piersy commented:
bors merge

piersy committed:
Log app startup error

piersy committed:
add more log

piersy committed:
Don't shut down services before app shutdown

piersy committed:
Prevent nip pointer exceptions

Remove some commented code

piersy committed:
Merge branch 'develop' into piersy/fix-app-startup

piersy committed:
Return error as last argument of startSynchronous

piersy committed:
Check for error before logging

piersy committed:
Merge branch 'develop' into piersy/fix-app-startup

piersy commented:
bors merge

piersy committed:
Will this fix the SIGSEGV

piersy committed:
Merge remote-tracking branch 'origin/develop' into piersy/fix-app-startup

piersy commented:
bors try

fasmat commented on .github/workflows/ci.yml:
```diff
@@ -194,6 +194,8 @@ jobs:
           GOTESTSUM_FORMAT: standard-verbose
           GOTESTSUM_JUNITFILE: unit-tests.xml
+          # We set this to prevent segmentation faults on macOS see - https://github.com/golang/go/issues/49138
+          MallocNanoZone: 0
```
This issue has been fixed in Go 1.17 if I understand the linked issue correctly, we are already using Go 1.20 🙂

piersy replied:
I can't find the run now, but I was seeing SIGSEV errors, I've removed it, lets see what happens.

fasmat commented on node/node.go:
```diff
@@ -167,6 +167,9 @@ func GetCommand() *cobra.Command {
 				// This blocks until the context is finished or until an error is produced
 				err = app.Start(ctx)
+				if err != nil {
+					app.log.With().Warning("App ended with error", log.Err(err))
```
This error is already logged on line 194

fasmat commented on node/node.go:
```diff
@@ -1413,46 +1448,36 @@ func (app *App) Start(ctx context.Context) error {
-		return nil
-	case err = <-app.errCh:
+	// cleanup shuts down components started here.
+	cleanup := func() error {
```
I feel like this should be part of `app.stopServices` instead of being returned via an anonymous function.

piersy replied:
Lets see how that looks

piersy replied:
Yep, that looks a lot nicer thanks :D

piersy committed:
Remove unneeded log statement

piersy committed:
Remove nano zone env var from ci

piersy committed:
Increase timeout for unit tests

The macOS unit test run timed out.

piersy committed:
Increase timeout for unit tests

The macos unit test job just took 28m to complete so a timeout of 30m seems
to be cutting it a bit fine. Increased to 35m.

piersy committed:
Do profiler and pprof shutdown in stopServices

Previously these services deferred their shutdown to the end of the
Start method.

fasmat commented on .github/workflows/ci.yml:
```diff
@@ -190,7 +190,7 @@ jobs:
         run: make clear-test-cache
       - name: unit tests
-        timeout-minutes: 25
+        timeout-minutes: 35
```
This increase shouldn't be necessary, unit tests take ~ 10 minutes on our self hosted runners and < 20 minutes on GH runners (except for Windows, but that should have a longer timeout already)

piersy replied:
It did take 28 minutes, in this run - https://github.com/spacemeshos/go-spacemesh/actions/runs/5966242390/job/16187625547. That's why I changed it.

I see it also taking 20 minutes in this run - https://github.com/spacemeshos/go-spacemesh/actions/runs/5967856416/job/16190433807.

piersy replied:
Hey @fasmat, in the most recent [run](https://github.com/spacemeshos/go-spacemesh/actions/runs/5968103737/job/16191204690) its back to 17 minutes, but it seems that it can fluctuate quite a lot. What do you want to do?

fasmat replied:
OK then leave the 35 minutes for now. I just want to avoid our CI becoming slower over time 🙂

fasmat commented on node/node.go:
```diff
@@ -1221,6 +1222,17 @@ func (app *App) stopServices(ctx context.Context) {
 	}
 
+	if app.pprofService != nil {
+		if err := app.db.Close(); err != nil {
```
```suggestion
		if err := app.pprofService.Close(); err != nil {
```

fasmat reviewed (commented):
lgtm, would like to hear @poszu opinion on the change too.

piersy committed:
Update node/node.go

correct copy paste error

Co-authored-by: Matthias Fasching <5011972+fasmat@users.noreply.github.com>

piersy committed:
Merge branch 'develop' into piersy/fix-app-startup

piersy committed:
Merge branch 'develop' into piersy/fix-app-startup

piersy commented:
This PR is now also suffering the same problem as #4897 see  https://github.com/spacemeshos/go-spacemesh/pull/4897#issuecomment-1693520633. My guess is that since these PRs both change the timing of shutdown events they are somehow triggering a previously unseen condition.

piersy committed:
Disable smeshing for tests in node_test.go

piersy commented:
bors merge

piersy committed:
Add fix for post verifier segfault on shutdown

piersy committed:
Revert "Disable smeshing for tests in node_test.go"

This reverts commit fe3d0f00b1cc7b02bca0875902784ad8b1b877f5.

piersy commented:
bors merge
